File: typescript/preliminar/src/app/common/middlewares/roles-permitted.middleware.ts
// Errors
import { UnauthorizedError } from "@/app/common/errors/unauthorized.error";

// Types
import {
  Request,
  Response,
  NextFunction,
} from "@/app/common/interfaces/http.interfaces";
import { RoleEnum } from "../enums/role.enum";

export const rolesPermittedMiddleware = (...rolesPermitted: RoleEnum[]) => {
  return (req: Request, res: Response, next: NextFunction) => {
    if (!req?.auth) {
      const error = new UnauthorizedError();
      return res.status(error.statusCode).json(error.toJSON());
    }

    if (!rolesPermitted.includes(req.auth.role)) {
      const error = new UnauthorizedError();
      return res.status(error.statusCode).json(error.toJSON());
    }

    next();
  };
};

File: go/fiber/internal/common/middleware/roles_permitted_middleware.go
package middleware

import (
	"tcc/internal/common/enum"
	"tcc/internal/common/errors"
	"tcc/internal/common/messages"
	auth "tcc/internal/modules/auth/interfaces"

	"github.com/gofiber/fiber/v2"
)

func RolesPermittedMiddleware(rolesPermitted ...enum.RoleEnum) fiber.Handler {
	return func(c *fiber.Ctx) error {
		jwtPayload := c.Locals("auth")
		if jwtPayload == nil {
			error := errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
			return c.Status(error.StatusCode).JSON(error)
		}

		payload, ok := jwtPayload.(auth.JwtPayload)
		if !ok || !contains(rolesPermitted, payload.Role) {
			error := errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
			return c.Status(error.StatusCode).JSON(error)
		}

		return c.Next()
	}
}

func contains(roles []enum.RoleEnum, role enum.RoleEnum) bool {
	for _, r := range roles {
		if r == role {
			return true
		}
	}
	return false
}
